### ISSUE TITLE
Restarting Runs with NetCDF Results in Problematic Re-Writing of the Initial Pressure File

### ISSUE THREAD

**Issue by nj1079** [parflow/parflow#552]
ParFlow Version: 3.13.0 (George's Derecho build) 

### **Context:** 
Running netCDF PF for runs of ~8760 timesteps (one water year). Sometimes, PF must be stopped and restarted in the middle of a water year (e.g. due to computational time limitations, to turn off seasonal settings, etc...). 

### **Issue:** 
When one restarts PF from a timestep, PF re-writes the initial pressure as a separate .nc file (of length 1). This causes problems when trying to process output because there is effectively an extra repeated timestep written out as output. All of these extra files could theoretically be deleted manually by the user before processing, but that seems unnecessary. 

### **Example:** 

Say I am running PF for one water year (8760 timesteps) and each .nc file contains up to 240 timesteps/file. I need to stop my run at timestep 5000 and restart it again at that same timestep and run to completion. 

In summary, I run from timestep 0 to timestep 5000. I then restart at timestep 5000 and run until 8760. 

PF will write output in 240-timestep increments up until 5000 (i.e. 00001, 00241, 00481, ...04801). Timestep 5000 occurs in the 04801 file. When I restart PF, the last timestep would be the last index of file 04801. This same timestep would also serve as the initial pressure for the continuation of the run. 

_**However, when restarting, PF will re-write a single separate file 05000 that is length one and is identical to the last timestep of the 04801 file (subtraction will equal zero). PF will then continue writing files every 240 timesteps as perscribed (i.e. 05001, 05241, ec...).**_ 

CLM does not do this; it's file outputs will skip right from 04801 to 05001 without re-printing the initial condition as a separate file. 

### **Proposed Fix:** 
It would be nice if ParFlow did not re-write an initial condition like this as a separate file in the .nc set up. It would be understandable if it re-wrote the last timestamp of 04801 as an initial condition (because they would be the same pressure anyways) if that is easier to fix, but re-writing a separate file is inconvenient. 

In the case of the example above, this would result in output numbering scheme consistent with that of CLM: 
04801  (with 200 timesteps) 
05001 (with 240 timesteps) 
_*note file 5000 does not exist!_ 

A related issue with .nc files is that the file numbering restarts every time you restart a run. For example, if you are printing 240 timesteps per file and you restart at timestep 803, you will output files will be numbered like this:  

00001 (with 240 timesteps)
00241 (with 240 timesteps)
00481 (with 240 timesteps)
00721 (with 83 timesteps)
00803 (with 240 timesteps)
00843(with 240 timesteps)
.....

It would be more convenient to read output if parflow could identify where within a .nc file it left off and then continue to print from there. In the example of restarting at timestep 803, files would be output with a consistent numbering pattern as follows: 
00001 (with 240 timesteps)
00241 (with 240 timesteps)
00481 (with 240 timesteps)
00721 (with 240 timesteps)
00961 (with 240 timesteps)
......

I only bring this up because this feature seems to be _**related**_ to the primary issue of re-writing the restart file as a separate file during restarts. That produces an extra timestep and is not intuitive if someone else is interpreting output. 

**_I am not suggesting that ParFlow be fixed to maintain consistent numbering when writing .nc files (though that would be nice!). I am suggesting that ParFlow be fixed to remove the re-write of the initial condition as a separate file during restarts._** 

I am happy to provide more context if needed! Thank you very much for your time.

**Comment by smithsg84** [comment]
Hi,

Can you share the Python or TCL script used to generate this behavior?

Thanks,

Steve

**Comment by nj1079** [comment]
Hi Steve,

I will attach my runscript to this email. However, if you want something
more condensed and easier for you to reproduce, please let me know and I'll
be *happy * to work on that for you! This runscript does a lot of stuff...

Best,
Nick

**Comment by smithsg84** [comment]
@nj1079 

It doesn't look like attachments work when responding to the issue emails.   You can attach a file to the using the github web interface and dragging and dropping into a comment or there is a button on the comment GUI to add files.

Thanks,
Steve

**Comment by nj1079** [comment]
Hi Steve, 

GitHub was giving me issues doing that for some weird reason. I just emailed you the script directly. 

-Nick

**Comment by smithsg84** [comment]
I'm not seeing an extra file being created in my tests.   What is the filename that has the single timestep?  The "file 05000" that you mention?

The script you sent is extracting the vector out of the NetCDF file and saving it to 'restart_pressure.pfb' and 'iniital_pressure.pfb' and looks to be using that for the restart.  Line 122-193 in the script.

**Comment by nj1079** [comment]
Hi Steve,

Thanks for your message. I apologize I typed this out yesterday and
realized that I forgot to click send.

If you are using my script to conduct runs, then that's something I'd be
happy to walk you though more if you'd like to make sure that the use case
is the same.

In this script, it would be the following case:

*Conducting a run starting like this:*

*runname* = <RUNNAME0007>

*istep              = 0* #start_time -->  (P)0-750 , (NoP)750-4748 ,
(P)4748-8760*stop_time          = 5000* if calendar.isleap(year) else
8760 # 8760 # 750 #start time + number simulation hours

*restart_directory* = <any restart directory>


Then, *after that run finishes*, then you would change the  istep to
5000, the stop_time to 8760/8784 (depending how many hours are in your
particular year), and change the restart directory to be the SAME
directory that you have just used/created in that run to run through
5000 timesteps:


*runname* = <RUNNAME0007>

*istep              = 5000* #start_time -->  (P)0-750 , (NoP)750-4748
, (P)4748-8760*stop_time          = 8760* if calendar.isleap(year)
else 8760 # 8760 # 750 #start time + number simulation hours

*restart_directory* = <RUNNAME0007>


This should restart from timetesp 5000 in the same directory as you
used to initially write timesteps 0 - 5000.


Please let me know if this does NOT produce the repeat initial file,
as that may mean there's some weird bug in my own version of the
runscript producing that behavior...



Let me know what you find, and thank you!


-Nick

**Comment by smithsg84** [comment]
I'm not running your script, I don't have all the files required.   When I run other PF runs I don't get a repeated file.

Your runscript creates a file with a single timestep, it is pulling the data for last timestep out of the netcdf file.   

```
    initial_pressure_restart_file = 'restart_pressure.pfb'
    print("initial pressure file is", initial_pressure_restart_file)
    cp(initial_pressure_restart_file,'initial_pressure.pfb')
    initial_file = 'initial_pressure.pfb'
```

And then uses this file for initial pressure.  Is the "repeat initial file" you are talking about these restart_pressure.pfb and initial_pressure.pfb files?  

I've looked over the files generated by the PF code and don't see were we would be writing out duplicated pressure files so not sure if I'm seeing the same issue you are.   I'm trying to figure out if you are seeing something PF is doing or your script.

**Comment by smithsg84** [comment]
I have looked over the netcdf code in PF and think I can add a feature to read a timestep from the netcdf file.   Currently the code is setup to read timestep 0 by default from netcdf files.   I believe it is a relatively easy fix to add another input key to specify the timestep.   Using a negative index like Python should make it easy to read the last timestep.    E.G. something like

```
UCRB.Geom.domain.ICPressure.FileName = initial_file
UCRB.Geom.domain.ICPressure.TimeStep = -1
```
Would grab the last timestep from the file.

**Comment by smithsg84** [comment]
Two new keys have been added to PF to help with the restarts.   NetCDF IC pressure files now support a TimeStep attribute to specify the timestep to read.   This can be an offset from the end.  To grab last timestep in a file use -1:

```
UCRB.Geom.domain.ICPressure.FileName = initial_file
UCRB.Geom.domain.ICPressure.TimeStep = -1
```

By default PF writes a set of files or in case of NetCDF a single file with multiple vectors when starting a run for the initial conditions and some unchanging values (like slopes).   The writing of the initial condition information can be disabled with using the Solver.PrintIntitialCondition key:

```
UCRB.Solver.PrintInitialConditions = False
```

**Comment by smithsg84** [comment]
Fixed in #558 and #559
